### PR TITLE
Export all functions to allow advanced optimizations.

### PR DESCRIPTION
**Pull request by dmjio**
https://developers.google.com/closure/compiler/docs/api-tutorial3#export

### CODE REVIEW

**Comment by dmjio** [comment]
cc @cocreature @FPtje @tysonzero @parthshah31, considering the following:

- [ ] Bumping nixpkgs pin to `ghcjs8.6`
- [ ] Refactoring `jsbits/*.js` to allow advanced optimizations
- [ ] Updating tests to reflect advanced optimizations changes.

Would appreciate a review, the `yarn` tests pass, and `todo-mvc` seems to pass a sanity test.

**Comment by dmjio** [comment]
```
------------------|----------|----------|----------|----------|-------------------|
File              |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------|----------|----------|----------|----------|-------------------|
All files         |    99.53 |    97.62 |      100 |    99.41 |                   |
 diff.cc.js       |    99.45 |    99.07 |      100 |     99.4 |               158 |
 isomorphic.cc.js |      100 |    88.89 |      100 |      100 |                 1 |
------------------|----------|----------|----------|----------|-------------------|
Test Suites: 1 passed, 1 total
Tests:       49 passed, 49 total
Snapshots:   0 total
Time:        4.635s
Ran all test suites.
Done in 5.22s.
```

File: default.nix [hunk 1]
```diff
@@ -1,32 +1,53 @@
...
     doHaddock = haddock;
     postInstall = ''
       mkdir -p $out/bin/mario.jsexe/imgs
       cp -r ${drv.src}/examples/mario/imgs $out/bin/mario.jsexe/
       cp ${drv.src}/examples/xhr/index.html $out/bin/xhr.jsexe/
-      ${closurecompiler}/bin/closure-compiler $out/bin/todo-mvc.jsexe/all.js > $out/bin/todo-mvc.jsexe/min.js
-      rm $out/bin/todo-mvc.jsexe/all.js
-      mv $out/bin/todo-mvc.jsexe/min.js $out/bin/todo-mvc.jsexe/all.js
+      ${closurecompiler}/bin/closure-compiler --compilation_level ADVANCED_OPTIMIZATIONS \
+        --jscomp_off=checkVars \
```
Comment (FPtje) [inline default.nix]: You might want to document this in e.g. the README. Which closure-compiler options should you run with and why?

File: tests/Main.hs [hunk 2]
```diff
@@ -95,7 +95,7 @@ propTest prop = do
     GaveUp { output = o } -> (False, o)
     Failure { output = o } -> (False, o)
     NoExpectedFailure { output = o } -> (False, o)
-    InsufficientCoverage { output = o } -> (False, o)
+--  InsufficientCoverage { output = o } -> (False, o)
```
Comment (FPtje) [inline tests/Main.hs]: Is commenting out the right thing, or should this code be removed?

File: default.nix (hunk 1 above)
Comment (dmjio) [reply default.nix]: yea, good call.

File: tests/Main.hs (hunk 2 above)
Comment (dmjio) [reply tests/Main.hs]: Probably removing altogether.

**Comment by dmjio** [comment]
@FPtje I think the issue with that error above is that the `foreign import`'s need to now be cognizant of the fact they're accessing the `window` object. I've pushed a commit to reconcile this fact, care to try with this hash? `757c31a`

File: default.nix (hunk 1 above)
Comment (dmjio) [reply default.nix]: But the GHCJS docs do this too, so maybe simply a link to this reference: https://github.com/ghcjs/ghcjs/wiki/Deployment

**Comment by FPtje** [comment]
Latest version confirmed working with isomorphic and jswidgets example, with modifications: I needed to string index all of my own FFI functions.

**Comment by FPtje** [comment]
Also confirmed working when not using (advanced) optimisations in both aforementioned examples.

**Comment by dmjio** [comment]
Examples pushed (`s3.sh`), confirmed working, and all tests pass for `diff.js` for both `ADVANCED` and omitted optimization.

**Pull request merged**